Parser/tokenizer.c

```diff
@@ -2437,18 +2437,45 @@ tok_get_fstring_mode(struct tok_state *tok, tokenizer_mode* current_tok, struct
     // If we start with a bracket, we defer to the normal mode as there is nothing for us to tokenize
     // before it.
     char start_char = tok_nextc(tok);
-    char peek = tok_nextc(tok);
-    tok_backup(tok, peek);
+    char peek1 = tok_nextc(tok);
+    char peek2 = tok_nextc(tok);
+    tok_backup(tok, peek2);
+    tok_backup(tok, peek1);
     tok_backup(tok, start_char);
 
-    if ((start_char == '{' && peek != '{') || (start_char == '}' && peek != '}')) {
+    if ((start_char == '{' && peek1 != '{') || (start_char == '}' && peek1 != '}')) {
         if (start_char == '{') {
             current_tok->bracket_mark[++current_tok->bracket_mark_index] = current_tok->bracket_stack;
         }
         tok->tok_mode_stack[tok->tok_mode_stack_index].kind = TOK_REGULAR_MODE;
         return tok_get_normal_mode(tok, current_tok, token);
     }
 
+    // Emit FSTRING_END in case we've reached the end of the string
+    if ((current_tok->f_string_quote_size == 1 && start_char == current_tok->f_string_quote)
+            || (current_tok->f_string_quote_size == 3 && start_char == current_tok->f_string_quote
+                && peek1 == current_tok->f_string_quote && peek2 == current_tok->f_string_quote)) {
```

Comment:
You can compare against `start_char` after `start_char == current_tok->f_string_quote` which is shorter. Also I think this whole thing can be simplified to:

```
    if (start_char == current_tok->f_string_quote && (current_tok->f_string_quote_size != 3  || (start_char ==  peek1 
 && start_char == peel2))) {
```